src/python/pants/engine/target.py

```diff
@@ -520,13 +520,41 @@ def types(self) -> Tuple[Type[Target], ...]:
         return tuple(self.aliases_to_types.values())
 
 
-# TODO: add light-weight runtime type checking to these helper fields, such as ensuring that
-# the raw_value of `BoolField` is in fact a `bool` and not an int or str. Use `instanceof`. All
-# the types are primitive objects like `str` and `int`, so this should be performant and easy to
-# implement.
-#
-# We should also sort where relevant, e.g. in `StringSequenceField`. This is important for cacahe
-# hits.
+class InvalidFieldException(Exception):
+    pass
+
+
+class InvalidFieldTypeException(InvalidFieldException):
+    """This is used to ensure that the field's value conforms with the expected type for the field,
+    e.g. `a boolean` or `a string` or `an iterable of strings and integers`."""
+
+    def __init__(
+        self, address: Address, field_alias: str, raw_value: Optional[Any], *, expected_type: str
+    ) -> None:
+        super().__init__(
+            f"The {repr(field_alias)} field in target {address} must be {expected_type}, but was "
+            f"`{repr(raw_value)}` with type `{type(raw_value).__name__}`."
+        )
+
+
+class UnimplementedField(PrimitiveField):
+    """Simply use whatever value the user passed in the BUILD file.
+
+    Warning: This is not safe to use with the engine. If the user uses an unhashable type like a
+    list, the engine will fail.
+    """
+
+    value: Optional[Any]
+
+    # TODO: maybe implement __hash__ on this so that this can always be used with the engine? For
```

Comment:
Worth raising an issue for higher visibility/discussions?

Reply:
Yes, but I think not yet. I plan to circle back to V1 backward-compatibility hopefully next week as one of the final parts of this project. There are some tricky questions there, such as how we support custom V1 target types. 

Likely, we're going to need to re-invent some sort of `TargetAdaptor` idea which has no V2 validation and simply allows us to pass the raw values down to V1.

This TODO ties in very closely to whatever comes out of that discussion.